Manta-Network/Manta
# feat: add ledger sync protocol

## Description

This PR adds the `pull` method for use by wallets to request the diff from the latest state.

### Unit Test Spec

Use `sudo` to set the storage to random UTXO and void numbers: for each round add random number (1 to 2) of random `(UTXO, EncryptedNote)` pairs and a random number (0 to 2) of random `VoidNumber`s:
- `Shards`: Add random `(UTXO, EncryptedNote)` pairs with a random first key and increasing second key
- `InsertionOrderedVoidNumbers`: Add random `VoidNumber`s with increasing key

Checkpoint Description:
- `receiver_index: [usize; 256]`: Lengths of each first key in `Shards`
- `sender_index: usize`: Length of `InsertionOrderedVoidNumber`

Testing correctness of Normal Usage:
1. Add random elements to ledger
2. Select a random checkpoint that's below the current one
3. Pull the latest state and check that it returns the data items (in order!) up to the checkpoint returned in the diff
4. If `should_continue == true` then `pull` should be called again with the returned checkpoint (GOTO step 3)
5. If it passes GOTO step 1

Testing correctness of Abnormal Usage: Same as the previous test except try to get from a checkpoint _later_ than the current one. The `pull` should return nothing.

### TODO

- [x] Implement diffing algorithm
- [x] Add RPC interface
- [ ] Add Unit tests / Simulation testing

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Updated relevant documentation in the code.
- [x] Added **one** line describing your change in `<branch>/CHANGELOG.md`
- [ ] If runtime changes, need to update the version numbers properly:
   * `authoring_version`: The version of the authorship interface. An authoring node will not attempt to author blocks unless this is equal to its native runtime.
   * `spec_version`: The version of the runtime specification. A full node will not attempt to use its native runtime in substitute for the on-chain Wasm runtime unless all of spec_name, spec_version, and authoring_version are the same between Wasm and native.
   * `impl_version`: The version of the implementation of the specification. Nodes are free to ignore this; it serves only as an indication that the code is different; as long as the other two versions are the same then while the actual code may be different, it is nonetheless required to do the same thing. Non-consensus-breaking optimizations are about the only changes that could be made which would result in only the impl_version changing.
   * `transaction_version`: The version of the extrinsics interface. This number must be updated in the following circumstances: extrinsic parameters (number, order, or types) have been changed; extrinsics or pallets have been removed; or the pallet order in the construct_runtime! macro or extrinsic order in a pallet has been changed. You can run the `metadata_diff.yml` workflow for help. If this number is updated, then the `spec_version` must also be updated 
- [x] Verify benchmarks & weights have been updated for any modified runtime logics
- [x] If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- [ ] If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- [ ] If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.
